### ISSUE TITLE
Bug: `TSESLint.FlatConfig.Config` isn't compatible with `Linter.FlatConfig`

### ISSUE THREAD

**Issue by alecmev** [typescript-eslint/typescript-eslint#8613]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### Repro Code

```TypeScript
import type { Linter } from "eslint";
import plugin from "typescript-eslint";

plugin.configs.disableTypeChecked satisfies Linter.FlatConfig;
```

### Expected Result

No error.

### Actual Result

```
... - error TS1360: Type 'Config' does not satisfy the expected type 'FlatConfig'.
  The types of 'languageOptions.parser' are incompatible between these types.
    Type 'LooseParserModule' is not assignable to type 'FlatConfigParserModule' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
      Type '{ meta?: Partial<ParserMeta>; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'FlatConfigParserModule' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
        Type '{ meta?: Partial<ParserMeta>; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Omit<ParserModule, "parseForESLint"> & { parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ...; }; }' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
          Type '{ meta?: Partial<ParserMeta>; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }; }'.
            The types returned by 'parseForESLint(...)' are incompatible between these types.
              Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }'.
                Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager">'.
                  Types of property 'visitorKeys' are incompatible.
                    Type 'unknown' is not assignable to type 'VisitorKeys | undefined'.

15 plugin.configs.disableTypeChecked satisfies Linter.FlatConfig;
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | `7.1.1` |
| `TypeScript`                           | `5.3.3` |
| `@types/eslint`                               | `8.56.5` |

**Comment by bradzacher** [comment]
This isn't surprising. We intentionally made our types loose so that the config could accept a wide variety of inputs for configs.

We first launched with super strict types but it turns out the ecosystem is messy as all hell and types re inconsistent at best.

Additionally we purposely do not seek compatibility with `@types/eslint` as well. Their types are not designed to take into account our AST and such.

We generally recommend using the ESLint classes and types we export from `@typescript-eslint/utils/ts-eslint` instead.

**Comment by alecmev** [comment]
I see, thanks! I saw https://github.com/typescript-eslint/typescript-eslint/issues/8467 and https://github.com/eslint/eslint/issues/18118, and thought the idea was to ultimately improve `@types/eslint`.

> We generally recommend using the ESLint classes and types we export from `@typescript-eslint/utils/ts-eslint` instead.

To clarify, what you're saying is that there's no reason to ever `import type ... from "eslint"` if there's access to `@typescript-eslint/utils/ts-eslint`, which is basically a better-maintained version of `@types/eslint`? If so, would it make sense to mention this in the docs? I couldn't to find anything relevant. Maybe some simple example:

```ts
import type { FlatConfig } from '@typescript-eslint/utils/ts-eslint';

export default [
  // ...
] satisfies FlatConfig.Config[];
```

And the rationale you used, "better interop with the ecosystem".

**Comment by bradzacher** [comment]
> To clarify, what you're saying is that there's no reason to ever import type ... from "eslint" if there's access to @typescript-eslint/utils/ts-eslint, which is basically a better-maintained version of @types/eslint?

Not better maintained, per se. Our jsdocs are better, sure, but in general the two should be roughly equivalent.
The big difference is the intent of the two packages. `@types/eslint` exists to allow you to add types that describe the entire state of the world - I.e. Are strictly correct. Our types are there to enable writing good, safe code with a good DevX.

> Maybe some simple example

Have you checked out the getting started guide as well as the `typescript-eslint` package docs which provides examples of the types for a flat config file and our recommended best practices for setting up flat configs?

https://typescript-eslint.io/getting-started/
https://typescript-eslint.io/packages/typescript-eslint#usage

The latter one, in particular, contains relevant examples.
